### PR TITLE
[Feature] Add a script to visualize lr

### PR DESCRIPTION
**Pull request by HAOCHENYE**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification
Add learning rate visualization script. The learning rate curve of target schedule registered in `EpochBasedRunner or IterBasedRunner` can be plot  as follows:

`python ./visualize_lr.py schedule_config.py `
The schedule_config.py can be:
```
lr_config = dict(
    policy='cyclic',
    # anneal_strategy='linear',
    target_ratio=(10, 1e-4),
    cyclic_times=5,
    step_ratio_up=0.4,
)

momentum_config = dict(
    policy='cyclic',
    # anneal_strategy='linear',
    target_ratio=(10, 1e-4),
    cyclic_times=5
)

runner = dict(type='IterBasedRunner',
              max_iters=1000)
```
The log of the runner will be saved in target work_dir and the learning rate curve will be plot according to the json log file:
image
The visualization of epoch based runner:
```
lr_config = dict(
    policy='cyclic',
    # anneal_strategy='linear',
    target_ratio=(10, 1e-4),
    cyclic_times=5,
    step_ratio_up=0.4,
)

momentum_config = dict(
    policy='cyclic',
    # anneal_strategy='linear',
    target_ratio=(10, 1e-4),
    cyclic_times=5
)
```
image

## BC-breaking (Optional)

No

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by zhouzaida** [comment]
We can also support the visualization of momentum.
https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/hooks/momentum_updater.py

File: .dev_scripts/visualize_lr.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+from torch.optim import SGD
+from torch.utils.data import DataLoader
+
+import mmcv
+from mmcv.runner import build_runner
+from mmcv.utils import get_logger
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='Train a detector')
```
Comment (zhouzaida) [inline .dev_scripts/visualize_lr.py:23]:
```suggestion
    parser = argparse.ArgumentParser(description='Visualize the given config of learning rate and momentum')
```

**Comment by mzr1996** [comment]
MMCLS also have learning rate visualization tool, which can be a reference.
https://mmclassification.readthedocs.io/en/latest/tools/visualization.html#learning-rate-schedule-visualization

**Comment by HAOCHENYE** [comment]
I have referenced the learning rate visualization of mmcls, and on this basis, this script can export the log and run with the training config, which is easy to check learning rate.

> MMCLS also have learning rate visualization tool, which can be a reference. https://mmclassification.readthedocs.io/en/latest/tools/visualization.html#learning-rate-schedule-visualization

**Review by teamwong111 — APPROVED** [review]

File: .dev_scripts/visualize_lr.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        assert isinstance(cur_lr, dict)
+        log_dict['lr'] = {}
+        for k, lr_ in cur_lr.items():
+            assert isinstance(lr_, list)
+            log_dict['lr'].update({k: lr_[0]})
+
+    if 'time' in runner.log_buffer.output:
+        # statistic memory
+        if torch.cuda.is_available():
+            log_dict['memory'] = self._get_max_memory(runner)
```
Comment (zhouzaida) [inline .dev_scripts/visualize_lr.py:106]: Is it necessary?

File: .dev_scripts/visualize_lr.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        self._iter += 1
+    self.call_hook('after_train_epoch')
+    self._epoch += 1
+
+
+def log(self, runner):
+    if 'eval_iter_num' in runner.log_buffer.output:
+        # this doesn't modify runner.iter and is regardless of by_epoch
+        cur_iter = runner.log_buffer.output.pop('eval_iter_num')
+    else:
```
Comment (zhouzaida) [inline .dev_scripts/visualize_lr.py:84]: We are visualizing the training process so can we remove those lines?

File: .dev_scripts/visualize_lr.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+            log = json.loads(line.strip())
+            data_dict['LearningRate'].append(log['lr'])
+            data_dict['Momentum'].append(log['momentum'])
+
+    wind_w, wind_h = [int(size) for size in cfg.window_size.split('*')]
+    # if legend is None, use {filename}_{key} as legend
+    fig, axes = plt.subplots(2, 1, figsize=(wind_w, wind_h))
+    plt.subplots_adjust(hspace=0.5)
+    font_size = 20
+    for index, (mode, data_list) in enumerate(data_dict.items()):
```
Comment (zhouzaida) [inline .dev_scripts/visualize_lr.py:182]: Suggest renaming `mode`

File: .dev_scripts/visualize_lr.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        if mode == 'LearningRate':
+            if cfg.get('lr_config'):
+                title = cfg.lr_config.type
+            else:
+                title = 'No learning rate schedule'
+        else:
+            if cfg.get('momentum_config'):
+                title = cfg.momentum_config.type
+            else:
+                title = 'No momentum schedule'
```
Comment (zhouzaida) [inline .dev_scripts/visualize_lr.py:210]:
```suggestion
                title = 'No momentum scheduler'
```

File: .dev_scripts/visualize_lr.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+            # plt.subplot(2, 1, index + 1)
+            x_list = np.arange(len(data_list)) * cfg.log_interval
+            ax.plot(x_list, data_list)
+            ax.set_xlabel('Iters', fontsize=font_size)
+        ax.set_ylabel(mode, fontsize=font_size)
+        if mode == 'LearningRate':
+            if cfg.get('lr_config'):
+                title = cfg.lr_config.type
+            else:
+                title = 'No learning rate schedule'
```
Comment (zhouzaida) [inline .dev_scripts/visualize_lr.py:205]:
```suggestion
                title = 'No learning rate scheduler'
```

**Review by zhouzaida — APPROVED** [review]
